## Section 1: Warm-up Questions


**Question 1.**

Suppose you are asked to develop a frontend program for a database system that backs an ecommerce website.
You need to write an `add_product` function, which allow people to add a new batch of products into the database.

The information people should provide at the point of call includes the product name, the quantity, the time when this batch is added (defaults to `None`), and, optionally, any other specifics (with keyword matching) about the batch to add.

Which of the following signatures is the most appropriate for this desired function:

- [ ] `add_product(name, quantity, time=None, *args)`
- [ ] `add_product(name=None, quantity, time=None, **kwargs)`
- [ ] `add_product(name=None, quantity=None, time=None, *args)`
- [x] `add_product(name, quantity, time=None, **kwargs)`




---

**Question 2.**

Given a function definition as follows, which of the following options is an incorrect way to call it?

```python
def print_all_args(x1, x2='3400', /, *args, y1,  y2='python', **kwargs):
    print("x1 is: ", x1)
    print("x2 is: ", x2)
    print("args is: ", args)
    print("y1 is: ", y1)
    print("y2 is: ", y2)
    print("kwargs is: ", kwargs)
```


- [ ] `print_all_args('3320', '2020', '3360', y1="java", title3="html")`
- [ ] `print_all_args('3320', title3="html", title4="CSS", y1='java')`
- [x] `print_all_args(y1="java", x1='3320', title3="html")`
- [ ] `print_all_args('3320', y1="java", title3="html")`


---


**Question 3.**

Read through the code snippet below, figure out what would be the outputs if the following inputs are provided and why:

- `32 // 0`

- `21 * 4`

- `34+3`

```python
while True:
    
    exp_str = input("Input an arithmetic expression (separate different parts with spaces): ")
    output = None
    
    try:
        
        str_ls = exp_str.split()
        operand2 = float(str_ls[2])  
        operand1 = float(str_ls[0])
    

        if str_ls[1] == '*':
            output = operand1 * operand2
        elif str_ls[1] == '/':
            output = operand1 / operand2    
        elif str_ls[1] == '+':
            output = operand1 + operand2
        elif str_ls[1] == '-':
            output = operand1 - operand2
               
        
    except IndexError:
        print("Failed to provide the right number of inputs. Remember to separate different parts with spaces!")
    
    except ValueError:        
        print("Failed to provide a valid numeric input.")
    
    except Exception as e:        
        print(f"Other errors: {e}")
    
    else:          
        if not output:
            print("Undefined operation!")
        else:   
            print(f"{' '.join(exp_str.split())} = {output}")
            break     
            
```    



The output of `32 // 0` is `Undefined operation!`. Explanation: [The operator // in str_ls[1] is not recognized, therefore output remains None.]

The output of `21 * 4` is `21 * 4 = 84.0`. Explanation: [No exceptions are raised, so the last else block executes.]

The output of `34+3` is `Failed to provide the right number of inputs. Remember to separate different parts with spaces!`. Explanation: [The input string is split into str_ls = ['34+3'] because there are no spaces separating the parts of the expression. Attempting to access str_ls[2] results in an IndexError, as the list does not have enough elements.]

---


## Section 2


For each question below, please write the corresponding code and then run the code to show the output.

---

**Question 1. Element-wise text concatenation**

Suppose you are given two lists, with their elements at the corresponding position representing the first name and the last name of the same person:

```python
first_names = ['Alice', 'Bob', 'Carol', 'David']
last_names = ['Alligator', 'Bear', 'Chimpanzee', 'Deer']
```

Write a lambda expression (hint: you need zip function and list comprehension), name it `concat_names`, and call it with the two lists above to form a list of full names. The expected output is as follows:


```python
['Alice Alligator', 'Bob Bear', 'Carol Chimpanzee', 'David Deer']
```

In [ ]:
first_names = ['Alice','Bob','Carol','David']
last_names = ['Alligator','Bear','Chimpanzee','Deer']

# Write your code below

concat_names = lambda fn, ln : [f'{first} {last}'for first, last in zip(fn,ln)]
concat_names(first_names, last_names)

['Alice Alligator', 'Bob Bear', 'Carol Chimpanzee', 'David Deer']

---


**Question 2. Use of `lambda`s as function arguments**



Suppose a music streaming service uses two lists to maintain a user's favorite songs and the corresponding ratings, respectively:

```python
songs = ['Back to Black', 'Poker Face', 'Yellow', 'Lolipop', 'All Too Well', 'Delicate', 'Moves Like Jagger']
ratings = [9.5, 8, 9, 8, 10, 7.5, 8]
```

Write code **without using loop statements** to create a playlist, which orders the user's favorite songs based on their ratings. And if a tie is observed, break the tie by sorting song titles lexicographically.


**Expected output**:


```python
[('All Too Well', 10),
 ('Back to Black', 9.5),
 ('Yellow', 9),
 ('Lolipop', 8),
 ('Moves Like Jagger', 8),
 ('Poker Face', 8),
 ('Delicate', 7.5)]
```


Tips: use `zip()` and `sorted()` with a lambda expression.

In [ ]:
songs = ['Back to Black', 'Poker Face', 'Yellow', 'Lolipop',
         'All Too Well', 'Delicate', 'Moves Like Jagger']

ratings = [9.5, 8, 9, 8, 10, 7.5, 8]

# Write your code below
sorted(zip(songs, ratings), key = lambda x : (-x[1],x[0]))

[('All Too Well', 10),
 ('Back to Black', 9.5),
 ('Yellow', 9),
 ('Lolipop', 8),
 ('Moves Like Jagger', 8),
 ('Poker Face', 8),
 ('Delicate', 7.5)]

---

**Question 3**


Suppose the organizer of the Hong Kong International Film Festival wants to determine the winner of a popularity award by conducting an online survey. In the survey form, respondents are asked to rate each nominated movie on a 5-point scale. The ratings collected by each form is stored as a flat Python list (e.g., `[4, 5, 3, 4, 5]` if there are 5 nominees), and the winner will be the movie that receives the highest average rating (or the largest sum of ratings).

Please define a custom Python function, `determine_winner(*forms)`, to perform the necessary calculation and return the index of the winner in the form (e.g., if the 2nd movie gets the highest average rating, return `2` as the output of this function). Hint: Given a list of numbers, we find the position of the maximum using `list.index()` and `max()`

For example, calling it with the data collected by the following 4 forms:

```python
collected_forms = [[4, 5, 3, 4, 5], [3, 5, 2, 4, 4], [4, 5, 4, 4, 5], [3, 4, 4, 3, 3]]
determine_winner(*collected_forms)
```

produces `2`.


Assume that there will be no ties in the case of mass responses, and no missing values as well (so every list fed into the function contains the same number of ratings).


In [ ]:
# Write your code below
def determine_winner(*forms):
  sum_ratings = [sum(ratings) for ratings in zip(*forms)]
  return sum_ratings.index(max(sum_ratings)) + 1

collected_forms = [[4, 5, 3, 4, 5], [3, 5, 2, 4, 4], [4, 5, 4, 4, 5], [3, 4, 4, 3, 3]]
determine_winner(*collected_forms)

2


---

**Question 4**

Many programming languages implement a function called `sum()`, which returns the sum of an arbitrary sequence of numbers.

Define a function `mySum` that allows us to sum up any number of numeric values. For example:

- We should be able to call `mySum(1, 12, 89, 5)` with 4 arguments or `mySum(*range(100))` with 100 arguments.
- Calling `mySum()` with no arguments returns zero.

The built-in function `sum()` is not allowed to use in the function body.

In [ ]:
# Write your function definition below
def mySum(*nums):
    sum = 0
    for num in nums:
        sum += num
    return sum

In [ ]:
mySum(1, 12, 89, 5)

107

In [ ]:
mySum(*range(100))

4950

---

**Question 5**

Parallel processing of multiple sequences is a powerful functionality for data analysis.

Define a function `mySumParallel(seqs)` that applies `mySum()` defined in **Question 4** to an arbitrary collection of sequences of numeric values in parallel. This function returns a list of sums of all sequences in the passed-in collection.

For example, calling `mySumParallel(collection)` where `collection = [[1, 12, 89, 5], range(100), range(2, 9, 3)]` returns `[107, 4950, 15]`.



In [ ]:
collection = [[1, 12, 89, 5], range(100), range(2, 9, 3)]

# Write your function definition below

def mySumParallel(seqs):
    return [mySum(*num_list) for num_list in seqs]


In [ ]:
mySumParallel(collection)

[107, 4950, 15]

---

**Question 6**

Further define a function `myParallel(aggfunc, seqs)` that allows any aggregation function to be applied to an arbitrary collection of sequences of numeric values.

For example, calling `myParallel(mySum, collection)` yields `[107, 4950, 15]`, while calling `myParallel(max, collection)` yields `[89, 99, 8]`.

In [ ]:
# Write your function definition below
def myParallel(aggfunc, seqs):
  return [aggfunc(*num_list) for num_list in seqs]


In [ ]:
myParallel(mySum, collection)


[107, 4950, 15]

In [ ]:
myParallel(max, collection)

[89, 99, 8]

---

**Question 7**

Given a partial class definition below:

```python
class Gradebook:
    
    def __init__(self, course_name, records):
        """
        Required arguments:
        course_name: a string that represents the course name
        records: a dictionary with each entry representing a student score pair
        """
        self.records = {}
        self.course_name = course_name
        self.records |= records
        
    def remove(self, student):
        del self.records[student]    
    
    def __repr__(self):   
        # substitute your code for pass below
        pass
    
    def update(self, student, score):  
        # substitute your code below for pass below
        pass
    
    def sort(self):   
        # substitute your code for pass below
        pass      

    def average(self):   
        # substitute your code for pass below
        pass  
```        

Complete the definitions for the methods `__repr__(self)`, `update(self, student, score)`, `sort(self)`, and `average(self)`  for the `Gradebook` class so that invoking them will generate outputs as follows:

*Code cell 1:*
```python
gradebook1 = Gradebook('Python Programming',
                       {'Troy': 92, 'Calvin': 95, 'James': 89, 'Charles': 100, 'Bryn': 59, 'Alice': 95})
gradebook1
```
*Output 1:*
```
Troy -> 92
Calvin -> 95
James -> 89
Charles -> 100
Bryn -> 59
Alice -> 95
```
*Code cell 2:*
```python
gradebook1.update("Amy", 100)
```
*Output 2:*
```
A new student record (Amy -> 100) is added!
```
*Code cell 3:*
```python
gradebook1.update("Bryn", 62)
```
*Output 3:*
```
Bryn's score has been updated from 59 to 62!
```

*Code cell 4:*
```python
gradebook1.sort()
gradebook1
```
*Output 4:* Note that the recods are sorted first by score, and then by name in alphabetically.
```
Amy -> 100
Charles -> 100
Alice -> 95
Calvin -> 95
Troy -> 92
James -> 89
Bryn -> 62
```

*Code cell 5:*
```python
gradebook1.average()
```
*Output 5:*
```
The average of the course Python Programming is: 90.43
```

In [ ]:
class Gradebook:

    def __init__(self, course_name, records):
        """
        Required arguments:
        course_name: a string that represents the course name
        records: a dictionary with each entry representing a student score pair
        """
        self.records = {}
        self.course_name = course_name
        self.records |= records

    def remove(self, student):
        del self.records[student]

    def __repr__(self):
        # substitute your code for pass below
        pass

    def update(self, student, score):
        # substitute your code below for pass below
        pass

    def sort(self):
        # substitute your code for pass below
        pass

    def average(self):
        # substitute your code for pass below
        pass

In [ ]:
class Gradebook:

    def __init__(self, course_name, records):
        self.records = {}
        self.course_name = course_name
        self.records |= records

    def remove(self, student):
        del self.records[student]

    def __repr__(self):
        str_ls = []
        for student, score in self.records.items():
            str_ls.append(f"{student} -> {score}")
        return '\n'.join(str_ls)

    def update(self, student, score):
        if student not in self.records:
            print(f"A new student record ({student} -> {score}) is added!")
        else:
            print(f"{student}'s score has been updated from {self.records[student]} to {score}!")
        self.records[student] = score

    def sort(self):
        self.records = dict(sorted(self.records.items(), key=lambda x: (-x[1], x[0])))


    def average(self):
        average_score = sum(self.records.values())/len(self.records)
        print(f"The average of the course {self.course_name} is: {average_score:.2f}")

In [ ]:
gradebook1 = Gradebook('Python Programming',
                       {'Troy': 92, 'Calvin': 95, 'James': 89, 'Charles': 100, 'Bryn': 59, 'Alice': 95})
gradebook1

Troy -> 92
Calvin -> 95
James -> 89
Charles -> 100
Bryn -> 59
Alice -> 95

In [ ]:
gradebook1.update("Amy", 100)

A new student record (Amy -> 100) is added!


In [ ]:
gradebook1.update("Bryn", 62)

Bryn's score has been updated from 59 to 62!


In [ ]:
gradebook1.sort()
gradebook1

Amy -> 100
Charles -> 100
Alice -> 95
Calvin -> 95
Troy -> 92
James -> 89
Bryn -> 62

In [ ]:
gradebook1.average()

The average of the course Python Programming is: 90.43
